frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceBatchAggregator.java

```diff
@@ -0,0 +1,213 @@
+package org.pytorch.serve.wlm;
+
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.pytorch.serve.job.Job;
+import org.pytorch.serve.job.JobGroup;
+import org.pytorch.serve.util.messages.BaseModelRequest;
+import org.pytorch.serve.util.messages.ModelWorkerResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SequenceBatchAggregator extends BatchAggregator {
+
+    private static final Logger logger = LoggerFactory.getLogger(SequenceBatchAggregator.class);
+    private ExecutorService pollExecutors;
+    private LinkedBlockingDeque<String> eventJobGroupIds;
+    private LinkedBlockingDeque<Job> jobQueue;
+    private Thread eventDispatcher;
+    private AtomicBoolean isPollJobGroup;
+    private LinkedList<String> currentJobGroupIds;
+
+    public SequenceBatchAggregator(Model model) {
+        super(model);
+        this.currentJobGroupIds = new LinkedList<>();
+        this.pollExecutors = Executors.newFixedThreadPool(model.getBatchSize() + 1);
+        this.jobQueue = new LinkedBlockingDeque<>();
+        this.isPollJobGroup = new AtomicBoolean(false);
+        this.eventJobGroupIds = new LinkedBlockingDeque<>();
+        this.eventJobGroupIds.add("");
+        this.eventDispatcher = new Thread(new EventDispatcher());
+        this.eventDispatcher.start();
+    }
+
+    public void startEventDispatcher() {
+        this.eventDispatcher.start();
+    }
+
+    public void stopEventDispatcher() {
+        this.eventDispatcher.interrupt();
+    }
+
+    private void pollJobGroup() throws InterruptedException {
+        if (isPollJobGroup.getAndSet(true)) {
+            return;
+        }
+        LinkedHashSet<String> tmpJobGroups = new LinkedHashSet<>();
+        String jobGroupId = model.getPendingJobGroups().poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        if (jobGroupId != null) {
+            addJobGroup(jobGroupId);
+
+            int quota = model.getPendingJobGroups().size() / model.getMaxWorkers();
+            if (quota > 0 && model.getPendingJobGroups().size() > 0) {
+                model.getPendingJobGroups().drainTo(tmpJobGroups, quota);
+            }
+
+            for (String jGroupId : tmpJobGroups) {
+                addJobGroup(jGroupId);
+            }
+        }
+        isPollJobGroup.set(false);
+    }
+
+    private void pollInferJob() throws InterruptedException {
+        Job job = jobQueue.poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
```

Comment:
Even though we expect this to be used in real time application we should also cover the case where we do request batching. Thus we need consider the batch delay here. Would be good to reuse the code [here ](https://github.com/pytorch/serve/blob/c56a702ea5a4159a5e8dd9b52935549a238f695c/frontend/server/src/main/java/org/pytorch/serve/wlm/Model.java#L376) and generalize so we do not create duplication.

Reply:
I'm confused if batchDelay should be applied once the first job is fetched in sequenceBatcher. Please confirm

Reply:
I think folks are gonna chose a very low delay due to the real time nature of the targeted use cases but we should still support it so users have the choice. In comparison to continuous batching there are new inputs coming in so it makes sense to have the option for a delay here.